### PR TITLE
Initial support for running etcd under kubelet

### PR DESCRIPTION
**Pull request by justinsb**
Very MVP at the moment, but does work if we replace the systemd
initsystem

### CODE REVIEW

File: initsystem/kubelet/initsystem.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+	desiredConfig *apis.EtcdAdmConfig
+}
+
+// EnableAndStartService enables and starts the etcd service
+func (s *InitSystem) EnableAndStartService(service string) error {
+	cfg := s.desiredConfig
+
+	name := s.name(cfg)
+
+	// We have to precreate the data dir for kubelet
```
Comment (dlipovetsky) [inline initsystem/kubelet/initsystem.go:43]: Not sure I understand. Please expand. Thanks!

File: initsystem/kubelet/initsystem.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+// SetConfiguration sets the desired etcd configuration
+func (s *InitSystem) SetConfiguration(cfg *apis.EtcdAdmConfig) error {
+	s.desiredConfig = cfg
+	return nil
+}
+
+func (s *InitSystem) buildPod(name string, cfg *apis.EtcdAdmConfig) (*pod, error) {
+	env := make(map[string]string)
+
+	// TODO: This is not the best way to build the environment, but it lets us minimize the initial changes
```
Comment (dlipovetsky) [inline initsystem/kubelet/initsystem.go:114]: :+1:

File: initsystem/kubelet/initsystem.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+			if len(tokens) != 2 {
+				return nil, fmt.Errorf("cannot parse environment line %q", line)
+			}
+
+			env[tokens[0]] = tokens[1]
+		}
+	}
+
+	// TODO: Critical pod annotations
+	// TODO: Should we use the full/official kubernetes API libraries?
```
Comment (dlipovetsky) [inline initsystem/kubelet/initsystem.go:137]: I think this is fine. Pulling in dependencies in the future will be easy.

File: initsystem/kubelet/initsystem.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+	}
+
+	podSpec := &pod.Spec
+
+	container := container{
+		Name: "etcd",
+	}
+
+	container.SecurityContext.Privileged = true
+	container.Resources.Requests = map[string]string{
```
Comment (dlipovetsky) [inline initsystem/kubelet/initsystem.go:158]: (tangent: Will we want to make Requests user-configurable later?)

File: service/unit.go [hunk 5]
```diff
@@ -41,13 +45,27 @@ func WriteEnvironmentFile(cfg *apis.EtcdAdmConfig) error {
...
+	var b bytes.Buffer
+
+	if err := t.Execute(&b, cfg); err != nil {
+		return nil, fmt.Errorf("unable to apply the etcd environment: %s", err)
+	}
+	return b.Bytes(), nil
+}
+
 // WriteUnitFile writes etcd service unit file
+// TODO: Move to systemd-specific area
```
Comment (dlipovetsky) [inline service/unit.go:68]: :+1:

File: service/unit.go [hunk 6]
```diff
@@ -41,13 +45,27 @@ func WriteEnvironmentFile(cfg *apis.EtcdAdmConfig) error {
...
-	if err := t.Execute(f, cfg); err != nil {
-		return fmt.Errorf("unable to apply the etcd environment: %s", err)
+	if _, err := f.Write(b); err != nil {
+		return fmt.Errorf("error writing etcd environment file %s: %w", cfg.EnvironmentFile, err)
 	}
+
 	return nil
 }
 
+// BuildEnvironment returns the environment variables corresponding to the desired configuration
```
Comment (dlipovetsky) [inline service/unit.go:55]: Thanks for refactoring this

**Comment by dlipovetsky** [comment]
/hold

Put a hold, just in case if you want to update the comment about pre-creating the data dir. Feel free to remove whenever.

/lgtm

**Comment by justinsb** [comment]
/reopen

Don't want to lose track of this one!

**Comment by mrajashree** [comment]
hi @justinsb @dlipovetsky this looks like a great addition, and estimate on when it will be merged?

File: initsystem/kubelet/initsystem.go (hunk 1 above)
Comment (dlipovetsky) [reply initsystem/kubelet/initsystem.go:43]: Looks like this to ensure correct ownership

For xref, kubeadm uses `DirectoryOrCreate` for this volume, but the owner is `kubelet`, and permissions are `755`.

```
  volumes:
  - hostPath:
      path: /etc/kubernetes/pki/etcd
      type: DirectoryOrCreate
    name: etcd-certs
  - hostPath:
      path: /var/lib/etcd
      type: DirectoryOrCreate
    name: etcd-data
```

**Comment by justinsb** [comment]
This was incorporated in the other PR; closing.

/close

**Pull request closed without merging**